File: js/src/beginner/ransomNote.js
import { strict as assert } from 'node:assert'
import logger from '../utils.js'

/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
function canConstruct(ransomNote, magazine) {
  for (const char of ransomNote.split('')) {
    if (!magazine.includes(char)) {
      return false
    } else {
      magazine = magazine.replace(char, '')
    }
  }

  return true
}

export function test() {
  logger.log('Running tests for canConstruct')
  assert(canConstruct('aa', 'aa'))
  assert(!canConstruct('a', 'b'))
  assert(!canConstruct('aaa', 'bbb'))
  logger.log('All tests passed!')
}

File: rs/src/beginner/ransom_note.rs
fn can_construct(ransom_note: String, mut magazine: String) -> bool {
    for c in ransom_note.chars() {
        if let Some(idx) = magazine.find(c) {
            magazine.remove(idx);
        } else {
            return false;
        }
    }
    true
}

pub fn test() {
    assert!(can_construct(String::from("aa"), String::from("aa")));
    assert!(!can_construct(String::from("a"), String::from("b")));
    assert!(!can_construct(String::from("aaa"), String::from("aab")));
}
